File: ios/Wrappers/ConversationListParamsWrapper.swift
import Foundation
import XMTP

struct ConversationListParamsWrapper {
	let createdAfterNs: Int64?
	let createdBeforeNs: Int64?
	let lastActivityAfterNs: Int64?
	let lastActivityBeforeNs: Int64?
	let limit: Int?
	let consentStates: [ConsentState]?
	let orderBy: ConversationsOrderBy?

	static func conversationListParamsFromJson(_ paramsJson: String)
		-> ConversationListParamsWrapper
	{
		guard !paramsJson.isEmpty else {
			return ConversationListParamsWrapper(
				createdAfterNs: nil,
				createdBeforeNs: nil,
				lastActivityAfterNs: nil,
				lastActivityBeforeNs: nil,
				limit: nil,
				consentStates: nil,
				orderBy: nil
			)
		}

		let data = paramsJson.data(using: .utf8) ?? Data()
		let jsonOptions =
			(try? JSONSerialization.jsonObject(with: data, options: []))
			as? [String: Any] ?? [:]

		let createdAfterNs = jsonOptions["createdAfterNs"] as? Int64
		let createdBeforeNs = jsonOptions["createdBeforeNs"] as? Int64
		let lastActivityAfterNs = jsonOptions["lastActivityAfterNs"] as? Int64
		let lastActivityBeforeNs = jsonOptions["lastActivityBeforeNs"] as? Int64
		let limit = jsonOptions["limit"] as? Int

		var consentStates: [ConsentState]? = nil
		if let statesArray = jsonOptions["consentStates"] as? [String] {
			consentStates = statesArray.map { state in
				switch state.lowercased() {
				case "allowed":
					return .allowed
				case "denied":
					return .denied
				default:
					return .unknown
				}
			}
		}

		var orderBy: ConversationsOrderBy? = nil
		if let orderByString = jsonOptions["orderBy"] as? String {
			switch orderByString.lowercased() {
			case "created_at":
				orderBy = .createdAt
			case "last_activity":
				orderBy = .lastActivity
			default:
				orderBy = .lastActivity
			}
		}

		return ConversationListParamsWrapper(
			createdAfterNs: createdAfterNs,
			createdBeforeNs: createdBeforeNs,
			lastActivityAfterNs: lastActivityAfterNs,
			lastActivityBeforeNs: lastActivityBeforeNs,
			limit: limit,
			consentStates: consentStates,
			orderBy: orderBy
		)
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/ConversationListParamsWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.JsonParser
import org.xmtp.android.library.ConsentState
import org.xmtp.android.library.Conversations

class ConversationListParamsWrapper(
    val createdAfterNs: Long?,
    val createdBeforeNs: Long?,
    val lastActivityAfterNs: Long?,
    val lastActivityBeforeNs: Long?,
    val limit: Int?,
    val consentStates: List<ConsentState>?,
    val orderBy: Conversations.ListConversationsOrderBy?,
) {
    companion object {
        fun conversationListParamsFromJson(paramsJson: String): ConversationListParamsWrapper {
            if (paramsJson.isEmpty()) {
                return ConversationListParamsWrapper(
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                )
            }

            val jsonOptions = JsonParser.parseString(paramsJson).asJsonObject

            val createdAfterNs =
                if (jsonOptions.has("createdAfterNs")) {
                    jsonOptions.get("createdAfterNs").asLong
                } else {
                    null
                }

            val createdBeforeNs =
                if (jsonOptions.has("createdBeforeNs")) {
                    jsonOptions.get("createdBeforeNs").asLong
                } else {
                    null
                }

            val lastActivityAfterNs =
                if (jsonOptions.has("lastActivityAfterNs")) {
                    jsonOptions.get("lastActivityAfterNs").asLong
                } else {
                    null
                }

            val lastActivityBeforeNs =
                if (jsonOptions.has("lastActivityBeforeNs")) {
                    jsonOptions.get("lastActivityBeforeNs").asLong
                } else {
                    null
                }

            val limit =
                if (jsonOptions.has("limit")) {
                    jsonOptions.get("limit").asInt
                } else {
                    null
                }

            val consentStates =
                if (jsonOptions.has("consentStates")) {
                    val statesArray = jsonOptions.getAsJsonArray("consentStates")
                    statesArray.map { state ->
                        when (state.asString.lowercase()) {
                            "allowed" -> ConsentState.ALLOWED
                            "denied" -> ConsentState.DENIED
                            else -> ConsentState.UNKNOWN
                        }
                    }
                } else {
                    null
                }

            val orderBy =
                if (jsonOptions.has("orderBy")) {
                    when (jsonOptions.get("orderBy").asString.lowercase()) {
                        "created_at" -> Conversations.ListConversationsOrderBy.CREATED_AT
                        "last_activity" -> Conversations.ListConversationsOrderBy.LAST_ACTIVITY
                        else -> Conversations.ListConversationsOrderBy.LAST_ACTIVITY
                    }
                } else {
                    null
                }

            return ConversationListParamsWrapper(
                createdAfterNs,
                createdBeforeNs,
                lastActivityAfterNs,
                lastActivityBeforeNs,
                limit,
                consentStates,
                orderBy,
            )
        }
    }
}
